benchmark_controller.py

```diff
@@ -5,63 +5,212 @@
+            self.args.compiler_flags.replace(
+                " ",
+                "") +
+            '_' +
+            self.args.machine_type +
+            '_' +
+            self.args.benchmark_options.replace(
+                " ",
+                ""))
+        self.binary_name = re.sub("[^a-zA-Z0-9_]+", "", identity).lower()
+        self.report_name = identity + '.report'
+
+    def _build_complete_flags(self):
+        if self.compiler_model is not None and self.benchmark_model is not None \
+                and self.machine_model is not None:
+            complete_build_flags, complete_link_flags = self.compiler_model.main(
+                'DEFAULT')
+            m_complete_build_flags, m_complete_link_flags = self.machine_model.main()
+            b_complete_build_flags, b_complete_link_flags = self.benchmark_model.fetch_flags()
+
+            complete_build_flags = complete_build_flags + ' ' + m_complete_build_flags + \
+                ' ' + b_complete_build_flags + ' ' + self.args.compiler_flags
+
+            complete_link_flags = complete_link_flags + ' ' + m_complete_link_flags + \
+                ' ' + b_complete_link_flags + ' ' + self.args.link_flags
+
+            complete_build_flags, complete_link_flags = self.compiler_model.validate_flags(
+                complete_build_flags, complete_link_flags)
+
+            return complete_build_flags, complete_link_flags
+
+    def _output_logs(self, stdout, perf_results):
+        result_dir = self.args.benchmark_root + self.binary_name + '/results'
+        if stdout and not isinstance(stdout, str):
+            raise TypeError('stdout should be a string of bytes')
+        if perf_results and not isinstance(perf_results, dict):
+            raise TypeError('perf_results should be a dictionary')
+        if not os.path.isdir(result_dir):
+            raise TypeError('%s should be a directory' % result_dir)
+
+        with open(result_dir + '/benchmark_stdout.report', 'w') as stdout_d:
+            stdout_d.write(stdout)
+
+        if isinstance(perf_results, dict):
+            with open(result_dir + '/perf_parser_results.report', 'w') as perf_res_d:
+                yaml.dump(perf_results, perf_res_d, default_flow_style=False)
+        else:
+            with open(result_dir + '/perf_parser_results.report', 'w') as perf_res_d:
+                perf_res_d.write(perf_results)
+
     def _load_benchmark_model(self, benchmark_name):
-        mod = importlib.import_module('models.' + benchmark_name)
+        mod = importlib.import_module('models.benchmarks.' + benchmark_name)
         return mod.BenchmarkModelImplementation()
 
-    def _validate_benchmark_model(self, benchmark_name):
-        filename = 'models/' + benchmark_name + '.py'
+    def _load_machine_model(self, machine_type):
+        mod = importlib.import_module('models.machines.' + machine_type)
+        return mod.MachineModelImplementation()
+
+    def _validate_model(self, model_name, model_type):
```

Comment:
I would run this from both load benchmark and machine model functions (ie. other way around), so that you only have one Import error if block.